mir_eval/sonify.py

```diff
@@ -81,36 +82,56 @@ def time_frequency(gram, frequencies, times, fs, function=np.sin, length=None):
     Returns
     -------
     output : np.ndarray
-        synthetized version of the piano roll
+        synthesized version of the piano roll
 
     """
     # Default value for length
+    if times.ndim == 1:
+        # Convert to intervals
+        times = util.boundaries_to_intervals(times)
+
     if length is None:
-        length = int(times[-1]*fs)
+        length = int(times[-1, 1] * fs)
+
+    times, _ = util.adjust_intervals(times, t_max=length)
+
+    # Truncate times so that the shape matches gram
+    times = times[:gram.shape[1]]
 
     def _fast_synthesize(frequency):
-        """A faster (approximate) way to synthesize a signal
-            synthesize a few periods then repeat that signal
+        """A faster way to synthesize a signal.
+            Generate one cycle, and simulate arbitrary repetitions
+            using array indexing tricks.
         """
-        # Generate ten periods at this frequency
-        ten_periods = int(10*fs*(1./frequency))
-        short_signal = function(2*np.pi*np.arange(ten_periods)*frequency/fs)
-        # Repeat the signal until it's of the desired length
+        # Generate one period at this frequency
+        short_signal = function(np.linspace(0, 2 * np.pi,
+                                            num=fs/float(frequency),
+                                            endpoint=False))
```

Comment:
Using `linspace` like this results in the wrong timebase, resulting the frequencies being detuned.  Assuming an integer sampling index `n`, if you want amplitude of a signal of frequency `f` with sampling rate `fs`, you need to evaluate `function` at `2*np.pi*n*frequency/fs`.  This `linspace` will evaluate `function` at `2*np.pi*n/int(fs/frequency)`:
```Python
fs = 11
frequency = 1.9
print np.linspace(0, 2 * np.pi, num=fs/float(frequency), endpoint=False)
print 2*np.pi*np.arange(5)*frequency/fs
print 2*np.pi*np.arange(5)/int(fs/frequency)
```
```
[ 0.          1.25663706  2.51327412  3.76991118  5.02654825]
[ 0.          1.08527746  2.17055492  3.25583239  4.34110985]
[ 0.          1.25663706  2.51327412  3.76991118  5.02654825]
```
I'm not sure how to coerce `linspace` to do the right thing here.